feat(containers): concepts improvements (#4059)
scaleway/docs-content: serverless/containers/concepts.mdx

## Container

A container is a package of software that includes all dependencies: code, runtime, configuration, and system libraries so that it can run on any host system. Scaleway provides custom Docker images that are entirely handled for you in the cloud.

## Container Registry

> Container Registry is the place where your images are stored before being deployed, we recommend using Scaleway Container Registry for a better integration. [Migration guide](/serverless/containers/api-cli/migrate-external-image-to-scaleway-registry/).

Reviewer: ```suggestion
Container Registry is the place where your images are stored before being deployed, we recommend using Scaleway Container Registry for optimal integration. [Migration guide](/serverless/containers/api-cli/migrate-external-image-to-scaleway-registry/).
```

## CRON trigger

A CRON trigger is a mechanism used to automatically invoke a Serverless Function at a specific time on a recurring schedule.

It works similarly to a traditional Linux [cron job](https://en.wikipedia.org/wiki/Cron), using the `* * * * *` format, and uses the **UTC** time zone. Refer to our [cron schedules reference](/serverless/containers/reference-content/cron-schedules/) for more information.

## Custom domain

> By default a generated endpoint is assigned to your Serverless resource and custom domains allows you to use your own domain read [custom domain documentation](/serverless/containers/how-to/add-a-custom-domain-to-a-container).

Reviewer: ```suggestion
By default, a generated endpoint is assigned to your Serverless resource. Custom domains allows you to use your own domain - see our [custom domain documentation](/serverless/containers/how-to/add-a-custom-domain-to-a-container) for full details.
```

## Deployment

Some parameter changes require a new deployment of the container to take effect. The deployment happens without causing downtime, as traffic is switched to the newest version.

## Endpoint

> An endpoint is the URL generated to access your resource, it can be customized with custom domains.

Reviewer: ```suggestion
An endpoint is the URL generated to access your resource. It can be customized with [custom domains](#custom-domain).
```

## Environment variables

> Environment variables are key/value pairs injected in your container. They are useful to share informations such as a configuration, with your container. Some names are reserved. [See details about reserved names](/serverless/containers/reference-content/containers-limitations/#configuration-restrictions).

Reviewer: ```suggestion
Environment variables are key/value pairs injected in your container. They are useful to share information such as configurations with your container. Some names are reserved. [See details about reserved names](/serverless/containers/reference-content/containers-limitations/#configuration-restrictions).
```

## Ephemeral storage

In addition to vCPU and RAM, Serverless Containers also provide a storage volume for the duration of the task. This storage space allows you to hold the data retrieved by the job, and disappears once the execution is complete.

The maximum size of the ephemeral storage is tied to the allocated memory.

## GB-s

Unit used to measure the resource consumption of a container. It reflects the amount of memory consumed over time.

## gRPC

> gRPC is supported on Serverless Containers, it requires to enable http2 (`h2c`) protocol.

Reviewer: ```suggestion
gRPC is supported on Serverless Containers, as long as you have enabled http2 (`h2c`) protocol.
```

## Healthcheck

> To determine the status of a container, default healthcheck automatically checks if basic requirements are met to define the status as `ready`.

Reviewer: ```suggestion
To determine the status of a container, the default healthcheck automatically checks if basic requirements are met, to define the status as `ready`.
```

You can define custom healthcheck rules with a [specific endpoint via the Scaleway API](https://www.scaleway.com/en/developers/api/serverless-containers/#path-containers-create-a-new-container).

## Instance

A Serverless Container instance handles incoming requests based on factors like the request volume, min scale, and max scale parameters.

## JWT Token

> JWT (JSON Web Token) is an access token you can create from the console or API to enable an application to access your Private Container. [Find out how to secure a container](/serverless/containers/how-to/secure-a-container/#restrict-access-to-your-containers)

Reviewer: ```suggestion
JWT (JSON Web Token) is an access token you can create from the console or API to enable an application to access your private container. [Find out how to secure a container](/serverless/containers/how-to/secure-a-container/#restrict-access-to-your-containers).
```

## Load balancing

> Incoming request traffic is managed by the Serverless infrastructure. In some cases such as rapid traffic changes or load testing, depending max scale parameter the ressources are automatically allocated to manage the traffic.
> 

Reviewer 2: ```suggestion
The Serverless infrastructure manages incoming request traffic. In scenarios like sudden traffic spikes or load testing, resources are automatically scaled based on the max scale parameter to handle the load.

```

## Max scale

> This parameter sets the maximum number of container instances. Adjust it based on your container's traffic spikes, but you might also limit the max scale to manage costs effectively.

Reviewer: ```suggestion
This parameter sets the maximum number of container instances. You should adjust it based on your container's traffic spikes, keeping in mind that you may wish to limit the max scale to manage costs effectively.
```

> It's the maximum number of instances of your Container. Define this parameter according to the traffic spikes of your Container. In some cases you may want to limit the max scale for cost control reasons.

Reviewer 2: ```suggestion
This parameter sets the maximum number of container instances. Adjust it based on your container's traffic spikes, but you might also limit the max scale to manage costs effectively.
```

## Min scale

> Customizing min scale for Serverless can be useful to keep an instance warm and limit cold starts. It also have an impact on costs of your Serverless Container.

Reviewer 2: ```suggestion
Customizing the minimum scale for Serverless can help ensure that an instance remains pre-allocated and ready to handle requests, reducing delays associated with cold starts. However, this setting also impacts the costs of your Serverless Container.
```

A container's privacy policy defines whether a container may be invoked anonymously (**public**) or only via an authentication mechanism provided by the [Scaleway API](https://www.scaleway.com/en/developers/api/serverless-containers/#authentication) (**private**).

## Registry endpoint

The registry endpoint parameter is the resource linked to the container image used in your Serverless Container.

## Sandbox

A sandbox is an isolation area for your container. Serverless Containers offer two sandboxing environments:

> - **v2** - Recommended for faster cold starts. Can introduce some overhead on specifc worklows with consequent amount of *syscalls*.

Reviewer 3: ```suggestion
- **v2** - Recommended for faster cold starts. Can introduce some overhead on specific worklows with consequent amount of *syscalls*.
```

A container's privacy policy defines whether a container may be invoked anonymously (**public**) or only via an authentication mechanism provided by the [Scaleway API](https://www.scaleway.com/en/developers/api/serverless-containers/#authentication) (**private**).

## Registry endpoint

> Registry endpoint parameter is the ressource link to the container image used in the Serverless Container.

Reviewer 2: ```suggestion
The registry endpoint parameter is the resource linked to the container image used in your Serverless Container.
```

## Serverless Framework

Serverless.com (Serverless Framework) is a tool that allows you to deploy serverless applications without having to manage Serverless Container's API call. Write and deploy a YAML configuration file, everything else is handled automatically, even the image building.

## Serverless Job

> Serverless Jobs are close to Serverless Containers but it's more adapted to run long workloads, see [comparaison between Serverless products](/serverless/containers/reference-content/difference-jobs-functions-containers.mdx).

Reviewer 2: ```suggestion
Serverless Jobs are similar to Serverless Containers but are better suited for running longer workloads. See [the comparaison between Serverless products](/serverless/containers/reference-content/difference-jobs-functions-containers) for more information.
```

## Stateless application

A stateless application is a computer program that does not save client data between sessions. Data generated in one session is not saved for use in the next session with that client.

> To store data persistently, use products like Databases or Object Storage.

Reviewer: ```suggestion
To store data persistently, use products like [Managed Databases]/managed-databases/) or [Object Storage](/storage/object/).
```

## Status

> A Serverless Container can be:

Reviewer 2: ```suggestion
A Serverless Container can have the following statuses:
```

* **Ready**: your Serverless Container is operational to serve requests.

> * **Pending**: your ressource is under deployment.
> * **Error**: something failed in the deploymen process, [open troubleshooting documentation](/serverless/containers/troubleshooting/cannot-deploy-image.mdx).

Reviewer 2: ```suggestion
* **Pending**: your resource is under deployment.
* **Error**: something went wrong during the deployment process. [Check our troubleshooting documentation](/serverless/containers/troubleshooting/cannot-deploy-image) to solve the issue.
```

## Terraform

> Terraform is an infrastructure as code tool, [read terraform documentation of Serverless Containers](https://registry.terraform.io/providers/scaleway/scaleway/latest/docs/resources/container).

Reviewer 2: ```suggestion
Terraform is a tool for managing infrastructure using code. [Read the Terraform documentation for Serverless Containers](https://registry.terraform.io/providers/scaleway/scaleway/latest/docs/resources/container).
```
